### DISCUSSION TITLE
M2A relation items could not greater than 100

### DISCUSSION THREAD

**Discussion by jianliao** [directus/directus#15554] [Feature Requests]
### Describe the Bug

When you have a complex collection that contains mass relation items, you readByQuery with a limit is -1. you would not be able to get all the relation item data. 

### To Reproduce

1. Create a collection that has more than 100 relation items in each of the item.
2. Run query
    const { data } = await directus.items('pages').readByQuery({
      limit: -1,
      filter: siteToQueryFilter[process.env.SITE],
      fields: '*.*.*.*.*.*.*.*'
    });
3. Observe the response data, you will notice that no more than 100 relation item will be returned

### What version of Directus are you using?

9.8.0

### What version of Node.js are you using?

16

### What database are you using?

Postgres

### What browser are you using?

Chrome

### How are you deploying Directus?

running locally

**Comment by azrikahar** [comment]
Just to clarify, do you mean the `limit: -1` should also apply to the M2A field? I believe the current behavior should be by design, unless I'm misunderstanding the description here 🤔 

For `readByQuery` or via query parameters, the use of [`deep` param](https://docs.directus.io/reference/query/#deep) is needed to apply limit (or any other query params) to relational/nested fields.

Here's an example:

- With `limit=-1`

   The `m2a` field still only returns 100 items, because this limit applies to the parent, not the m2a field:
   
  ![chrome_Qgf5LriPg9](https://user-images.githubusercontent.com/42867097/163323354-658d335c-aaf5-4768-8e43-1345047264ca.png)
  
- Added `deep[m2a][_limit]=-1`

   Now the m2a field does get the limit -1 applied, and returns 101 items instead:

   (technically the original `limit=-1` is not needed here as it's not affecting the m2a field)
      
   ![chrome_9hw4MrSuU5](https://user-images.githubusercontent.com/42867097/163323545-d8e9067a-4643-4f5a-8562-c1ec9ab7e176.png)

**Comment by jianliao** [comment]
hmm...How do I apply `deep[m2a][_limit]=-1`? like this?
```
    const { data } = await directus.items('pages').readByQuery({
      limit: -1,
      filter: siteToQueryFilter[process.env.SITE],
      fields: '*.*.*.*.*.*.*.*',
      m2a: {
        _limit: -1
      }
    });
```
I tested it, but it doesn't work. Still see a lot of `null` items in response.

**Comment by jianliao** [comment]
@azrikahar I put a lot of detail in the description in #adobe private discord channel. Could you please join that channel? Thanks for the help. @w0kyj might help.

**Comment by rijkvanzanten** [comment]
> hmm...How do I apply `deep[m2a][_limit]=-1`? like this?
> 
> ```
>     const { data } = await directus.items('pages').readByQuery({
>       limit: -1,
>       filter: siteToQueryFilter[process.env.SITE],
>       fields: '*.*.*.*.*.*.*.*',
>       m2a: {
>         _limit: -1
>       }
>     });
> ```
> 
> I tested it, but it doesn't work. Still see a lot of `null` items in response.

The deep query would be

```js
    const { data } = await directus.items('pages').readByQuery({
      limit: -1,
      filter: siteToQueryFilter[process.env.SITE],
      fields: '*.*.*.*.*.*.*.*',
      deep: {
        m2a: {
          _limit: -1
        }
      }
    });
```

**Comment by jianliao** [comment]
Thanks @rijkvanzanten . I just tried it and still not working.

**Comment by jianliao** [comment]
I tried `a2o` as well. Still see the 100 limit.
```
    const { data } = await directus.items('pages').readByQuery({
      limit: 23,
      filter: siteToQueryFilter[process.env.SITE],
      fields: '*.*.*.*.*.*.*.*',
      deep: {
        a2o: {
          _limit: -1
        },
        m2a: {
          _limit: -1
        }
      }
    });
```

**Comment by rijkvanzanten** [comment]
`m2a` / `a2o` should be the name of your actual `m2a` field. Is your field called `m2a`? @azrikahar's example above uses a field called `m2a`, hence his filter using `deep[m2a]` 🙂

**Comment by azrikahar** [comment]
> @azrikahar I put a lot of detail in the description in #adobe private discord channel. Could you please join that channel? Thanks for the help. @w0kyj might help.

Will do! Thanks 👍

---

Was just going to ask about the field name but seems like Rijk is one step ahead as usual! 😄

**Comment by jianliao** [comment]
The field name is `item`. Is it correct?
```
    const { data } = await directus.items('pages').readByQuery({
      limit: 23,
      filter: siteToQueryFilter[process.env.SITE],
      fields: '*.*.*.*.*.*.*.*',
      deep: {
        a2o: {
          _limit: -1
        },
        m2a: {
          _limit: -1
        },
        item: {
          _limit: -1
        }
      }
    });
```
But still not working. I have 106 referenced items, only 100 got filled, so 6 is `null`.
![image](https://user-images.githubusercontent.com/1207520/163423858-77c76cfa-fcc6-4548-a894-dada171f504e.png)

**Comment by jianliao** [comment]
This item field is a little bit deep. Do I need to insert all the parent properties?

**Comment by rijkvanzanten** [comment]
> This item field is a little bit deep. Do I need to insert all the parent properties?

Yes! You might end up with something like:

```
deep
  sections
    page
      modules
        _limit: -1
```

depending on the field names / relationships used

**Comment by jianliao** [comment]
I'll try. But it is really not flexible. Every time I create a new data model, to return all the data, I would need to change my query code.

**Comment by jianliao** [comment]
Still does not work. I tried:
```
deep: {
        sections: {
          section: {
            content: {
              item: {
                _limit: -1
              }
            }
          }
        }
      }
```

and

```
deep: {
        page: {
          sections: {
            section: {
              content: {
                item: {
                  _limit: -1
                }
              }
            }
          }
        }
   }
```
A simplified response looks like this:
```
{
  "sections": [
    {
      "id": 22273,
      "sort": 3,
      "section": {
        "id": "b4d56b9d-b986-4529-b068-79f9e65dc4d1",
        "title": "Usage guidelines",
        "text": null,
        "visibility": "all",
        "display_title": true,
        "content": [
          {
            "id": 41930,
            "sort": 0,
            "collection": "usage_guideline",
            "item": null
          }
        ]
      }
    }
  ]
}
```

**Comment by jianliao** [comment]
Could you please take a look at my PR #12749? It is just passing in the query limit from the context.

**Comment by jianliao** [comment]
Tried this as well, still no luck.
```
deep: {
        // a2o: {
        //   _limit: -1
        // },
        // m2a: {
        //   _limit: -1
        // },
        // item: {
        //   _limit: -1
        // },
        sections: [
          {
            section: {
              content: [{
                item: {
                  _limit: -1
                }
              }]
            }
          }
        ]
      }
```

**Comment by rijkvanzanten** [comment]
I believe the list you want to limit on is `content`, not the `item` within. Can you try

```
deep: {
        page: {
          sections: {
            section: {
              content: {
                _limit: -1
              }
            }
          }
        }
   }
```

**Comment by jianliao** [comment]
Tried it, but still no luck.

**Comment by emrahnazif** [comment]
In my setup with v9.5.2

`deep={"XXX": {"_limit": YYY}}`

works fine. But there might be a problem with getting relational data with v9.6.0 and above versions of directus. I had a quick chance to test it and it was not working. As I didn't validate the issue, I didn't open a ticket but @jianliao if it is possible you may try with v9.5.2 to see if it will make any difference. 

A+

**Comment by jianliao** [comment]
Thanks, @emrahnazif. But the problem is that I don't like the way to configure the limit. It is too verbose, my data model is very complex and has frequent changes. I don't know when the data will reach the limit(100), I don't want to configure this deep parameter every time I change the data model. I am currently on 9.4.3, yes I can try upgrading to 9.5.2, but the fix I need is in 9.8.0+.

**Comment by ta-le** [comment]
If someone is still looking for a quick solution to this, you can apply a patch to change this default limit from 100 to some larger number:

https://github.com/directus/directus/blob/a803331b4aeaa5c027fe5e847129ad005cbba818/api/src/database/run-ast.ts#L246

This fixes the issue where querying deeply nested content returned `null` values after exceeding the limit.

**Comment by Nitwel** [comment]
This issue seems to have turned into a discussion about how we can improve setting the limit on relational data rather than an issue with with the `deep` operator itself. I would love to discuss how we can improve the current implementation.

**Comment by rdsubhas** [comment]
Hi, we have the same issue here. Many deeply relational models, and specifying this one-by-one for each nested field is really difficult.

Especially, when we were migrating from v8 to v9, and this caught us by surprise. https://docs.directus.io/self-hosted/upgrades-migrations.html => even though it's not listed in breaking changes docs, we were wondering why half the stuff were not showing up. quite big behavior change even though it looks subtle.

In HTTP conventions, the limit for the URL path is 2048 characters. Adding many deep limits can cross this size.

I kindly ask to consider some alternatives – for ergonomics, for technical limits (2048 http URL) and for avoiding migration bugs:

1. if nested limit is not specified, instead of defaulting to 100 – use the top-level `?limit` from the API call please?
2. or to have a top-level limit setting for deep, like `deep[_limit]=-1` or `[deep][*][_limit]=-1`
3. or to have a project setting in directus for the default limit (we were relying on v8 Project-level setting, and part of the breaking change is that, if this default limit was provided as an option it could have been reused)

Thank you!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
